### ISSUE TITLE
[BUG] The maximum configured depth of

### ISSUE THREAD

**Issue by shravan2x** [microsoft/playwright-dotnet#2781]
### System info
- Playwright Version: v1.40.0
- Operating System: Windows 11
- Browser: Chromium
- Other info:

### Source code

- [ ] I provided exact source code that allows reproducing the issue locally.

**Steps**
- Go to any webpage
- `JsonElement someResult = await _page.MainFrame.EvaluateAsync<JsonElement>("<script to generate recursive objects with depth 100 or 1000>");`
- `someResult.Deserialize<MyType>(new JsonSerializerOptions() { MaxDepth = 256 })!;`

**Expected**

It doesn't seem practical to allow infinite depth, so the ideal solution seems to be to allow developers to configure the JSON serializer defaults.

**Actual**

```
Microsoft.Playwright.TargetClosedException: The maximum configured depth of 64 has been exceeded. Cannot read next JSON object. Path: $.result | LineNumber: 0 | BytePositionInLine: 25123.
```

**Comment by shravan2x** [comment]
I tried serializing my data to a string before returning, but websites can modify their default JS functions by force. So client-side JS functions like `JSON.stringify` aren't reliable.

**Comment by mxschmitt** [comment]
Would it be possible to provide a full reproducible so we can run it locally?

**Comment by mxschmitt** [comment]
What is your use-case? We recommend doing JSON.stringify instead for now as a workaround, since Chromium does not allow more than 100:

     [0104/154443.812125:FATAL:devtools_session.cc(408)] Check failed: status.ok(). CBOR: stack limit exceeded at position 3635

Something like:

```csharp
var someResult = await page.EvaluateAsync<string>(@"() => {
    let result = { name: 'foo' };
    for (let i = 0; i < 65; i++) {
        result = { child: result, name: `child-${i}` };
    }
    return JSON.stringify(result);
}");
```

**Comment by shravan2x** [comment]
That was our first approach, however, some websites override the built-in `JSON.stringify` function leading to malformed JSON. Without a way to "reset" the client's JSON method, my only option is to send the raw data back via Playwright.

**Comment by mxschmitt** [comment]
You could do something like this, its in Node.js, but would work in .NET as well:

```ts
it('', async ({ page }) => {
  await page.context().addInitScript(() => {
    globalThis[Symbol.for('JsonStringify')] = JSON.stringify;
  });
  await page.goto('https://example.com');
  await page.evaluate(() => {
    JSON.stringify = null;
  });
  const value = await page.evaluate(() => {
    let result = { name: 'foo', child: null };
    for (let i = 0; i < 65; i++)
      result = { child: result, name: `child-${i}` };
    return globalThis[Symbol.for('JsonStringify')](result);
  });
  console.log(value);
});
```

Have you tried that approach?

**Comment by shravan2x** [comment]
Yeah. The problem is that [Javascript allows users to set custom overrides for JSON serialization](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#:~:text=If%20the%20value%20has%20a%20toJSON()%20method%2C%20it%27s%20responsible%20to%20define%20what%20data%20will%20be%20serialized.%20Instead%20of%20the%20object%20being%20serialized%2C%20the%20value%20returned%20by%20the%20toJSON()%20method%20when%20called%20will%20be%20serialized.) via `toJSON` methods.

Even if I managed to save the original definition of `JSON.stringify`, saving+restoring the `toJSON` methods of objects, arrays, numbers, strings, booleans, dates, etc is difficult and error prone.

**Comment by mxschmitt** [comment]
Well, yeah we in our upstream serializer traverse the object and ignore the `toJSON` property. Let me file it upstream, since it affects microsoft/playwright and decide there how we proceed.

Do you mind sharing what kind of circular data structure your are serializing? So we can decide better what kind of use-case this is in order to priotize it. Thanks!
